homeassistant/helpers/event.py

```diff
@@ -90,34 +99,246 @@ def state_change_listener(event):
 track_state_change = threaded_listener_factory(async_track_state_change)
 
 
+_boolean_coerce = vol.Any(cv.boolean, lambda val: False)
+
+
 @callback
 @bind_hass
-def async_track_template(hass, template, action, variables=None):
-    """Add a listener that track state changes with template condition."""
-    from . import condition
+def async_track_template(
+        hass: HomeAssistant,
+        template: Template,
+        action: Callable[[str, Optional[State], Optional[State]], None],
+        variables: Optional[TemplateVarsType] = None) \
+        -> Callable[[], None]:
+    """Add a listener that fires when a a template evaluates to 'true'.
+
+    Listen for the result of the template becomming true, or a true-like
+    string result, such as 'On', 'Open', or 'Yes'. If the template results
+    in an error state when the value changes, this will be logged and not
+    passed through.
+
+    On template registration, the action will be called with an entity_id
+    and state of some entity referenced by the template, or '*' if no
+    entity is referenced by the template.
+
+    If the initial run of the template is invalid and results in an
+    exception, the listener will still be registered but will only
+    fire if the template result becomes true without an exception.
+
+    Action arguments
+    ----------------
+    entity_id
+        ID of the entity that triggered the state change.
+    old_state
+        The old state of the entity that changed.
+    new_state
+        New state of the entity that changed.
+
+    Parameters
+    ----------
+    hass
+        Home assistant object.
+    template
+        The template to calculate.
+    action
+        Callable to call with results. See above for arguments.
+    variables
+        Variables to pass to the template.
+
+    Returns
+    -------
+    Callable to unregister the listener.
+
+    """
+    @callback
+    def state_changed_listener(event, template, last_result, result):
+        """Check if condition is correct and run action."""
+        if isinstance(result, TemplateError):
+            _LOGGER.exception(result)
+            return
+
+        if last_result is not None:
+            if not _boolean_coerce(last_result) \
+                    and _boolean_coerce(result):
+                hass.async_run_job(action, event.data.get('entity_id'),
+                                   event.data.get('old_state'),
+                                   event.data.get('new_state'))
+        elif _boolean_coerce(result):
+            # First run of the listener. Figure out an entity ID to
+            # pass back to the action because it expects one.
```

Comment:
We should pass the entity ID whose state change caused us to re-evaluate the template and realize it was changed.

Reply:
This happens only with the very first fire of the template, so there's no specific entity associated with the fire. This code makes a guess about what it might be and uses that. It's the same outcome as what used to happen so avoids any breaking change.

If anything `None` would be more appropriate, since there's no particular entity, but that would break things that use it.